File: src/formats/lrc.ts
// LRC (.lrc) lyrics adapter — compatible with the sibling Lyrical Sync app.
//
// Each lyric line is "[mm:ss.xx] text". A cue's end = the next line's start
// (the LRC format has no explicit end); the last line gets a default tail.
// Metadata tags ([ar:], [ti:], [offset:] …) are ignored on parse. Multi-line
// cue text is joined with spaces on export (LRC lines are single-line).

import { newCueId, emptyDocument, type Cue, type SubtitleDocument } from "../types/subtitle";
import { sortedCues } from "./srt";

const LAST_TAIL = 4; // seconds of display for the final line
const TIME_TAG = /\[(\d{1,2}):(\d{1,2})(?:[.:](\d{1,3}))?\]/g;

export function parseLrc(raw: string): SubtitleDocument {
  const doc = emptyDocument("lrc");
  const rows: Array<{ start: number; text: string }> = [];
  for (const line of raw.replace(/\r\n/g, "\n").split("\n")) {
    TIME_TAG.lastIndex = 0;
    const tags: number[] = [];
    let m: RegExpExecArray | null;
    while ((m = TIME_TAG.exec(line)) !== null) {
      const frac = m[3] ? Number((m[3] + "00").slice(0, 2)) / 100 : 0;
      tags.push(Number(m[1]) * 60 + Number(m[2]) + frac);
    }
    if (!tags.length) continue; // metadata / blank line
    const text = line.replace(TIME_TAG, "").trim();
    for (const start of tags) rows.push({ start, text }); // one tag may repeat a line
  }
  rows.sort((a, b) => a.start - b.start);
  rows.forEach((row, i) => {
    const end = i + 1 < rows.length ? rows[i + 1].start : row.start + LAST_TAIL;
    doc.cues.push({ id: newCueId(), start: row.start, end, text: row.text });
  });
  return doc;
}

function lrcTime(seconds: number): string {
  const s = Math.max(0, seconds);
  const mm = Math.floor(s / 60);
  const ss = Math.floor(s % 60);
  const cc = Math.round((s - Math.floor(s)) * 100);
  const pad = (n: number) => String(n).padStart(2, "0");
  return `[${pad(mm)}:${pad(ss)}.${pad(cc >= 100 ? 99 : cc)}]`;
}

export function serializeLrc(doc: SubtitleDocument): string {
  const out = sortedCues(doc.cues).map((cue: Cue) => `${lrcTime(cue.start)}${cue.text.replace(/\n/g, " ")}`);
  return out.join("\n") + "\n";
}

File: native/Sources/GlyphlineCore/Formats/LRC.swift
// LRC (.lrc) lyrics adapter (ported from ../../src/formats/lrc.ts).
//
// Each lyric line is "[mm:ss.xx] text". A cue's end = the next line's start; the
// last line gets a default tail. Metadata tags ([ar:], [ti:], …) are ignored.
// Multi-line text joins with spaces on export (LRC lines are single-line).

import Foundation

private let LRC_LAST_TAIL = 4.0
private let lrcTagRegex = try! NSRegularExpression(pattern: #"\[(\d{1,2}):(\d{1,2})(?:[.:](\d{1,3}))?\]"#)

public func parseLrc(_ raw: String) -> SubtitleDocument {
    var doc = SubtitleDocument.empty(.lrc)
    var rows: [(start: Double, text: String)] = []
    for line in raw.replacingOccurrences(of: "\r\n", with: "\n").components(separatedBy: "\n") {
        let ns = line as NSString
        let matches = lrcTagRegex.matches(in: line, range: NSRange(location: 0, length: ns.length))
        if matches.isEmpty { continue } // metadata / blank
        var tags: [Double] = []
        for m in matches {
            let mm = Double(ns.substring(with: m.range(at: 1))) ?? 0
            let ss = Double(ns.substring(with: m.range(at: 2))) ?? 0
            let fracRange = m.range(at: 3)
            let frac = fracRange.location == NSNotFound
                ? 0.0
                : (Double(String((ns.substring(with: fracRange) + "00").prefix(2))) ?? 0) / 100
            tags.append(mm * 60 + ss + frac)
        }
        let text = lrcTagRegex.stringByReplacingMatches(
            in: line, range: NSRange(location: 0, length: ns.length), withTemplate: ""
        ).trimmed()
        for start in tags { rows.append((start, text)) }
    }
    rows.sort { $0.start < $1.start }
    for (i, row) in rows.enumerated() {
        let end = i + 1 < rows.count ? rows[i + 1].start : row.start + LRC_LAST_TAIL
        doc.cues.append(Cue(id: newCueId(), start: row.start, end: end, text: row.text))
    }
    return doc
}

private func lrcTime(_ seconds: Double) -> String {
    let s = max(0, seconds)
    let mm = Int(s / 60)
    let ss = Int(s.truncatingRemainder(dividingBy: 60))
    var cc = Int(((s - s.rounded(.down)) * 100).rounded())
    if cc >= 100 { cc = 99 }
    func p(_ n: Int) -> String { n < 10 ? "0\(n)" : "\(n)" }
    return "[\(p(mm)):\(p(ss)).\(p(cc))]"
}

public func serializeLrc(_ doc: SubtitleDocument) -> String {
    let lines = sortedCues(doc.cues).map { cue in
        lrcTime(cue.start) + cue.text.replacingOccurrences(of: "\n", with: " ")
    }
    return lines.joined(separator: "\n") + "\n"
}
